librz/debug/dsession.c

```diff
@@ -63,6 +63,10 @@ RZ_API bool rz_debug_add_checkpoint(RzDebug *dbg) {
 	rz_debug_reg_sync (dbg, RZ_REG_TYPE_ALL, 0);
 	for (i = 0; i < RZ_REG_TYPE_LAST; i++) {
 		RzRegArena *a = dbg->reg->regset[i].arena;
+		// Ignore empty register arenas
+		if (!a->bytes) {
```

Comment:
How can this happen? I see that there is a `rz_reg_arena_new` function and that can't return an arena with bytes == null. I think these kind of things shall be caught much before.

Reply:
When [`rz_reg_fit_arena`](https://github.com/rizinorg/rizin/blob/dev/librz/reg/arena.c#L134) is called after [register profile creation](https://github.com/rizinorg/rizin/blob/dev/librz/reg/profile.c#L237) the size is recalculated and freed if it's smaller than 1:
https://github.com/rizinorg/rizin/blob/4f51dd53514a4ae8ee7d694629ca3a2e3ca15fae/librz/reg/arena.c#L146-L153

and since [`rz_debug_reg_sync`](https://github.com/rizinorg/rizin/blob/dev/librz/debug/dreg.c#L8) receives size 0 when performing `reg_read` from the debug backend the arena will remain null:

https://github.com/rizinorg/rizin/blob/4f51dd53514a4ae8ee7d694629ca3a2e3ca15fae/librz/debug/dreg.c#L69-L72

> rz_reg_arena_new function and that can't return an arena with bytes == null

As I wrote in the description, that's what caused the segfault since b->size was defaulted to 1. I am in favor of making that new() return null for size 0 but I am not sure if something else might break. Do we want to look into it in this PR?

Reply:
i think this is the way to go. i would still have a safe memcpy define or inline which we can use to also avoid this stuff..

Reply:
IMO there are some invariants that we should try to maintain to make the code easy to understand and avoid all these `if`/`else` scattered everywhere. For example, an arena is created in `rz_reg_arena_new` with the assumption that, if it is created, `->arena` is allocated and it is at least `->size` bytes. Now, this condition should *always* be respected IMHO, so you don't have to check this everywhere. I think it makes sense to try and see if removing that `size < 1` check in `rz_reg_arena_new` works.